### ISSUE TITLE
`:root_source_code_paths` compile time vs runtime

### ISSUE THREAD

**Issue by pierot** [getsentry/sentry-elixir#614]
### Environment

9.0

### Steps to Reproduce

1. I followed the documentation for adding `enable_source_code_context` together with `root_source_code_paths`.
2. In our CI we have a compile and build step.
3. The compile time fetches deps and compiles them.
4. Then our tests run and after they pass we build in a Docker container the project with the _build , deps and mix.lock output of the compile step. That way we can leverage caching.

### Expected Result

Successful compile.

### Actual Result

```
ERROR! the application :sentry has a different value set for key :root_source_code_paths during runtime compared to compile time. Since this application environment entry was marked as compile time, this difference can lead to different behaviour than expected:
  * Compile time value was set to: ["/builds/XXX/YYY/ZZZ"]
  * Runtime value was set to: ["/opt/app"]
To fix this error, you might:
  * Make the runtime value match the compile time one
  * Recompile your project. If the misconfigured application is a dependency, you may need to run "mix deps.compile sentry --force"
  * Alternatively, you can disable this check. If you are using releases, you can set :validate_compile_env to false in your release configuration. If you are using Mix to start your system, you can pass the --no-validate-compile-env flag
```

I sort of get what the error says. But I can't figure out how to solve this.
While building the `File.cwd!()` was `/builds/....` and when running the `mix phx.digest` in our build step, it is in the `/opt/app`.

**Comment by Cervajz** [comment]
Hello,

we have the same issue after updating to 9.0. We did not perform any other change than the update. 

```
ERROR! the application :sentry has a different value set for key :enable_source_code_context during runtime compared to compile time. Since this application environment entry was marked as compile time, this difference can lead to different behaviour than expected

* Runtime value was set to: true 
* Compile time value was not set 
```

Our configuration is not different between compilations. We use static `true` for that key:

```elixir
config :sentry,
  dsn: configuration.sentry.dsn_key,
  environment_name: String.to_atom(configuration.app.environment_name),
  enable_source_code_context: true,
  root_source_code_paths: [File.cwd!()],
  included_environments: [:canary_staging, :staging, :canary_production, :production],
  report_deps: true
```

we compile the application inside Docker:

```Dockerfile
FROM hexpm/elixir:1.14.5-erlang-26.1-alpine-3.18.2 AS build

ENV LANG=C.UTF-8 \
    MIX_ENV=prod

RUN apk add --no-cache build-base libstdc++ git python3 nodejs=18.17.1-r0 yarn=1.22.19-r0

# elixir dependencies
RUN mix do local.hex --force, local.rebar --force

WORKDIR /app

COPY config ./config/
COPY mix.exs mix.lock ./

ARG OBAN_KEY_FINGERPRINT
ARG OBAN_LICENSE_KEY

# Oban Web+Pro
RUN mix hex.repo add oban https://getoban.pro/repo  \
      --fetch-public-key $OBAN_KEY_FINGERPRINT \
      --auth-key $OBAN_LICENSE_KEY

RUN mix do deps.get --only=$MIX_ENV, deps.compile
RUN mix esbuild.install

# Dashboard npm dependencies
COPY assets/package.json ./assets/
COPY assets/yarn.lock ./assets/

RUN yarn --cwd ./assets install --frozen-lockfile --silent

# esbuild and elixir release
COPY assets ./assets/
COPY lib ./lib/
COPY priv ./priv
COPY rel ./rel/

ARG GIT_REVISION_HASH
ARG GIT_BRANCH_SLUG

# See config/config.exs for more details.
ENV WEB_SECRET_KEY_BASE="nope"

ENV DASHBOARD_REPO_DATABASE_URL="nope"
ENV API_REPO_DATABASE_URL="nope"

ENV GIT_REVISION_HASH=$GIT_REVISION_HASH
ENV GIT_BRANCH_SLUG=$GIT_BRANCH_SLUG

RUN mix compile && \
    mix assets.deploy && \
    mix sentry_recompile && \
    mix release ek

FROM hexpm/elixir:1.14.5-erlang-26.1-alpine-3.18.2 AS app

# See all available and mandatory environment variables in config/runtime.exs
ENV LANG=C.UTF-8 \
    HOME=/app

RUN apk add --no-cache openssl ncurses-libs curl bash libstdc++ nmap htop

WORKDIR /app

RUN chown nobody:nobody /app

USER nobody:nobody

COPY --from=build --chown=nobody:nobody "/app/_build/prod/rel/ek" ./

CMD ["bin/ek", "start"]
```

**Comment by warmwaffles** [comment]
I just ran into this as well.

**Comment by warmwaffles** [comment]
Actually just figured it out. The following config is what we had

```elixir
config :sentry,
  # ... snip ...
  enable_source_code_context: true,
  root_source_code_path: File.cwd!(),
  # ... snip ...
```

Notice the singular `root_source_code_path`. I swapped that to be `root_source_code_paths: [File.cwd!()],` and it seems to work.

**Comment by krainboltgreene** [comment]
Yes, this does appear to be a documentation mismatch/changelog issue: In this comment it adds the alternative `_paths`, but the documentation has always referred to `_path`. This comment fully removes it: https://github.com/getsentry/sentry-elixir/commit/8d9d42d712e2b90fe6bebb57f96fc7ad6ef20ae8

It just never made it to soft deprecation *or* in the changelog.

**Comment by whatyouhide** [comment]
@Cervajz see this piece of docs: https://hexdocs.pm/elixir/Config.html#module-config-runtime-exs

You cannot configure compile-time configuration in `config/runtime.exs` because it is evaluated _after_ the code is compiled. Can you move the compile-time config to `config/config.exs` (or `config/dev.exs`, or whatever you use) and see if that fixes it? See https://hexdocs.pm/sentry/upgrade-9-x.html#fix-your-environment-variables also.

**Comment by Cervajz** [comment]
@whatyouhide Thank you for the upgrade guide, that definitely helped!. From my side, the issue is resolved 🎉

**Comment by whatyouhide** [comment]
@pierot where are you configuring `:root_source_code_paths`?

**Comment by pierot** [comment]
> But let's try to address it! My main question is this: do you run mix phx.digest after building (mix deps.compile and mix compile)? Why don't you run them together?

This is in our Dockerfile

```
RUN source .env.$DEPLOY_ENV && \
    mix deps.get --only $BUILD_ENV & \
    yarn install --ignore-optional --prefer-offline --pure-lockfile && \
    wait

RUN source .env.$DEPLOY_ENV && \
    make _build_production_js & make _build_production_styles && \
    wait && \
    mix phx.digest
```

> @pierot where are you configuring `:root_source_code_paths`?

We configure this in `config/config.exs`.

**Comment by whatyouhide** [comment]
@pierot yeah, this is not an issue with the configuration I think. You're building the `sentry` application in a context, and then you're building things again in another context where the config value is different. Try this: when you build the Docker image, can you do `mix deps.compile --force sentry` before other things? That should recompile `sentry` with the right new compile-time config. Let me know if that works!

**Comment by pierot** [comment]
@whatyouhide Yeah, I know this is the real issue. But that means we can't leverage pre-compile caching which is than used for our test jobs and eventuallay (when they succeed) for our final build. I wil try and see what `mix deps.compile --force sentry` gives us.

**Comment by whatyouhide** [comment]
`mix deps.compile --force sentry` should not recompile everything to be clear: it will only recompile Sentry (and maybe parts of your code), so it shouldn't introduce too much overhead.

**Comment by bautrey37** [comment]
We're running into this issue and we've tried the things mentioned in the thread. We've done the fixes as described the docs [here](https://docs.sentry.io/platforms/elixir/?original_referrer=https%3A%2F%2Fwww.google.com%2F#including-source-code). Are these up to date?

**Comment by pierot** [comment]
Error remains. We've reverted the setting and are not using 
```elixir
config :sentry,
  # ... snip ...
  enable_source_code_context: true,
  root_source_code_path: File.cwd!(),
  # ... snip ...
```
anymore.

**Comment by whatyouhide** [comment]
> We're running into this issue and we've tried the things mentioned in the thread. We've done the fixes as described the docs [here](https://docs.sentry.io/platforms/elixir/?original_referrer=https%3A%2F%2Fwww.google.com%2F#including-source-code). Are these up to date?

@bautrey37 No, that documentation is not up to date. I'm gonna do a pass on it in October.

> Error remains

@pierot I'd love a minimal reproducible setup that looks somewhat like your Docker-based setup in order to debug this, since it gets hard pretty quickly here on the GitHub issue. I’m surprised that doing `mix deps.compile sentry --force` right before other operations once you move to the build step doesn't fix this.

**Comment by guess** [comment]
I'm working in an umbrella question and using `File.cwd!()` makes it difficult because I have to run different mix commands within the specific apps (e.g., `mix ecto.setup` in /apps/db or `mix assets.deploy` in /apps/web).

To solve, I am doing this in my config.exs:
```elixir
project_root = fn ->
  root = "/app"

  case String.split(File.cwd!(), root) do
    [root | _] -> root <> root
    _ -> root
  end
end

...

config :sentry,
  ...
  enable_source_code_context: true,
  root_source_code_paths: [project_root.()]
```

Where `/app` is just where all the code lives in my docker container:
```
WORKDIR /app
```

But I have a question for umbrella apps:
Should the list of `root_source_code_paths` just be the root of the umbrella project? Or is it expected to individually specify all the app paths?

**Comment by whatyouhide** [comment]
This was closed by #632. The `:root_source_code_paths` configuration is not a compile-time configuration anymore.
